ddtrace/internal/ci_visibility/_api_client.py

```diff
@@ -0,0 +1,524 @@
+
+    record_skippable_count(len(tests_to_skip), TEST)
+
+    return tests_to_skip
+
+
+class _TestVisibilityAPIClientBase(abc.ABC):
+    """Client for fetching test visibility settings from the CI Visibility API
+
+    This class makes no direct references to environment variables, configs, or settings not passed to its
+    constructor (except for default values).
+    """
+
+    _requests_mode: REQUESTS_MODE
+
+    def __init__(
+        self,
+        base_url: str,
+        itr_skipping_level: ITR_SKIPPING_LEVEL,
+        git_data: GitData,
+        configurations: t.Dict[str, t.Any],
+        dd_service: t.Optional[str] = None,
+        dd_env: t.Optional[str] = None,
+        timeout: t.Optional[float] = None,
+    ):
+        self._base_url: str = base_url
+        self._itr_skipping_level: ITR_SKIPPING_LEVEL = itr_skipping_level
+        self._git_data: GitData = git_data
+        self._configurations: _CONFIGURATIONS_TYPE = configurations
+        self._service: t.Optional[str] = dd_service
+        self._dd_env: t.Optional[str] = dd_env
+        self._timeout: float = timeout if timeout is not None else DEFAULT_TIMEOUT
+
+    @abc.abstractmethod
+    def _redact_headers(self) -> t.Dict[str, str]:
+        """This is an abstract method to force child classes to consider which headers should be redacted for logging"""
+        pass
+
+    @abc.abstractmethod
+    def _get_headers(self) -> t.Dict[str, str]:
+        pass
+
+    def _get_final_headers(self) -> t.Dict[str, str]:
+        headers = _BASE_HEADERS.copy()
+        headers.update(self._get_headers())
+        return headers
+
+    def _do_request(self, method: str, endpoint: str, payload: str, timeout: t.Optional[float] = None) -> Response:
+        timeout = timeout if timeout is not None else self._timeout
+        headers = self._get_final_headers()
+        url = combine_url_path(self._base_url, endpoint)
+
+        conn: t.Optional[t.Union[HTTPSConnection | HTTPConnection | UDSHTTPConnection]] = None
+        try:
+            parsed_url = verify_url(url)
+            url_path = parsed_url.path
+            conn = get_connection(url, timeout)
+
+            log.debug(
+                "Sending %s request: %s %s %s %s",
+                self._requests_mode.name,
+                method,
+                url,
+                payload,
+                self._redact_headers(),
+            )
+
+            conn.request("POST", url_path, payload, headers)
+            resp = conn.getresponse()
+            log.debug("Response status: %s", resp.status)
+            response = Response.from_http_response(resp)
+            return response
+        finally:
+            if conn is not None:
+                conn.close()
+
+    def fetch_settings(self) -> TestVisibilityAPISettings:
+        """Fetches settings from the test visibility API endpoint
+
+        This raises encountered exceptions because fetch_settings may be used multiple times during a session.
```

Comment:
I did not get why raising exceptions is useful because the function may be used multiple times :thinking: